components/suggest/src/metrics.rs

```diff
@@ -32,39 +32,48 @@ pub struct SuggestIngestionMetrics {
 impl SuggestIngestionMetrics {
     /// Wraps each iteration in `ingest` and records the time for it.
     ///
-    /// Passes the closure a DownloadTimer.  Use this to measure the times for all
-    /// downloads that happen during the ingest
+    /// Passes the closure a `&mut MetricsContext`.
     pub fn measure_ingest<F, T>(&mut self, record_type: impl Into<String>, operation: F) -> T
     where
-        F: FnOnce(&mut DownloadTimer) -> T,
+        F: FnOnce(&mut MetricsContext) -> T,
     {
         let timer = Instant::now();
         let record_type = record_type.into();
-        let mut download_metrics = DownloadTimer::default();
-        let result = operation(&mut download_metrics);
+        let mut context = MetricsContext::default();
+        let result = operation(&mut context);
+        if !context.processed_record {
+            // Don't record metrics if we didn't actually processed anything.
+            return result;
+        }
         let elapsed = timer.elapsed().as_micros() as u64;
         self.ingestion_times.push(LabeledTimingSample::new(
             record_type.clone(),
-            elapsed - download_metrics.total_time,
-        ));
-        self.download_times.push(LabeledTimingSample::new(
-            record_type,
-            download_metrics.total_time,
+            elapsed - context.total_time,
         ));
+        self.download_times
+            .push(LabeledTimingSample::new(record_type, context.total_time));
         result
     }
 }
 
-/// Records download times for a single loop in ingest
-///
-/// [Self::measure_download] can be called multiple times.  [DownloadTimer] will track the total
-/// time for all calls.
+/// Context for a ingestion measurement
 #[derive(Default)]
-pub struct DownloadTimer {
+pub struct MetricsContext {
+    processed_record: bool,
```

Comment:
Suggestion/nit: the consumers having to call `will_process_record()` is a little annoying. How about using an enum to represent the context?

```
enum Context {
    Uninstrumented,
    Instrumented {total_time: u64},
}

pub struct MetricsContext {
    context: Context,
}

// The `context` defaults to `Uninstrumentated` and later can shift to `Instrumented` whenever `measure_download()` gets called.
```